rpcs3/Emu/Cell/lv2/sys_usbd.h

```diff
@@ -1,64 +1,185 @@
-#pragma once
+﻿#pragma once
 
+#include <queue>
+#include <libusb.h>
 #include "Emu/Memory/vm_ptr.h"
+#include "Emu/Io/usb_device.h"
 
-// SysCalls
+#define MAX_SYS_USBD_TRANSFERS 0x44
 
-struct DeviceListUnknownDataType
+// PS3 internal codes
+enum PS3StandardUsbErrors : u32
 {
-	u8 unk1;
-	u8 deviceID;
-	u8 unk3;
-	u8 unk4;
-} ;
-
-/* 0x01 device descriptor */
-struct deviceDescriptor
+	HC_CC_NOERR     = 0x00,
+	EHCI_CC_MISSMF  = 0x10,
+	EHCI_CC_XACT    = 0x20,
+	EHCI_CC_BABBLE  = 0x30,
+	EHCI_CC_DATABUF = 0x40,
+	EHCI_CC_HALTED  = 0x50,
+};
+
+enum PS3IsochronousUsbErrors : u8
+{
+	USBD_HC_CC_NOERR   = 0x00,
+	USBD_HC_CC_MISSMF  = 0x01,
+	USBD_HC_CC_XACT    = 0x02,
+	USBD_HC_CC_BABBLE  = 0x04,
+	USBD_HC_CC_DATABUF = 0x08,
+};
+
+enum SysUsbdEvents : u32
+{
+	SYS_USBD_ATTACH            = 0x01,
+	SYS_USBD_DETACH            = 0x02,
+	SYS_USBD_TRANSFER_COMPLETE = 0x03,
+	SYS_USBD_TERMINATE         = 0x04,
+};
+
+// PS3 internal structures
+struct UsbInternalDevice
+{
+	u8 device_high; // System flag maybe (used in generating actual device number)
+	u8 device_low;  // Just a number identifying the device (used in generating actual device number)
+	u8 unk3;        // ? Seems to always be 2?
+	u8 unk4;        // ?
+};
+
+struct UsbDeviceRequest
 {
-	u8 bLength;              /* descriptor size in bytes */
-	u8 bDescriptorType;      /* constant USB_DESCRIPTOR_TYPE_DEVICE */
-	u16 bcdUSB;              /* USB spec release compliance number */
-	u8 bDeviceClass;         /* class code */
-	u8 bDeviceSubClass;      /* subclass code */
-	u8 bDeviceProtocol;      /* protocol code */
-	u8 bMaxPacketSize0;      /* max packet size for endpoint 0 */
-	u16 idVendor;            /* USB-IF Vendor ID (VID) */
-	u16 idProduct;           /* Product ID (PID) */
-	u16 bcdDevice;           /* device release number */
-	u8 iManufacturer;        /* manufacturer string descriptor index */
-	u8 iProduct;             /* product string desccriptor index */
-	u8 iSerialNumber;        /* serial number string descriptor index */
-	u8 bNumConfigurations;   /* number of configurations */
+	u8 bmRequestType;
+	u8 bRequest;
+	be_t<u16> wValue;
+	be_t<u16> wIndex;
+	be_t<u16> wLength;
+};
+
+struct UsbDeviceIsoRequest
+{
+	vm::ptr<void> buf;
+	be_t<u32> start_frame;
+	be_t<u32> num_packets;
+	be_t<u16> packets[8];
+};
+
+// HLE usbd
+void LIBUSB_CALL callback_transfer(struct libusb_transfer* transfer);
+
+struct UsbTransfer
+{
+	u32 transfer_id;
```

Comment:
uninitialized

Reply:
They're all Initialized at the creation of the usb thread but fine I'll add = 0.